src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java

```diff
@@ -0,0 +1,260 @@
+
+            swerveModules[2] = new SwerveModule(2, DRIVE_MOTOR_BACK_RIGHT, ANGLE_MOTOR_BACK_RIGHT, BACK_RIGHT_INVERTED,
+                    Constants.SwerveModule.SICK_ANGLE_PIDF, Constants.SwerveModule.DRIVE_PIDF);
+
+            swerveModules[3] = new SwerveModule(3, DRIVE_MOTOR_BACK_LEFT, ANGLE_MOTOR_BACK_LEFT, BACK_LEFT_INVERTED,
+                    Constants.SwerveModule.ANGLE_PIDF, Constants.SwerveModule.DRIVE_PIDF);
+        }
+
+        this.isFieldOriented = isFieldOriented;
+    }
+
+    /**
+     * Sets the wheels of the robot to the calculated angle and speed.
+     *
+     * @param forward  the Y value of the joystick
+     * @param strafe   the X value of the joystick
+     * @param rotation the rotation Z of the joystick
+     */
+    public void holonomicDrive(double forward, double strafe, double rotation) {
+
+        double[] robotHeading = getRobotHeading(strafe, forward, rotation, -Math.toRadians(Robot.gyro.getAngle()));
+
+        double[] velocities = calculateWheelVelocities(robotHeading);
+        double[] polar;
+        double[][] controls = new double[4][2];
+
+        // converts the cartesian velocities to polar and transfers them to a control matrix
+        for (int i = 0; i < 4; i++) {
+            polar = Utils.cartesianToPolar(velocities[2 * i + 1], velocities[2 * i]);
+            controls[i][0] = polar[0];
+            controls[i][1] = polar[1];
+        }
+
+        // feeds the corresponding control to each wheel
+        for (int k = 0; k < 4; k++) {
+            swerveModules[k].setSpeed(controls[k][0]);
+            swerveModules[k].setAngle(controls[k][1]);
+        }
+
+        double sumx = 0;
+        double sumy = 0;
+        for (int j = 0; j < 4; j++) {
+            sumx += velocities[j * 2];
+            sumy += velocities[j * 2 + 1];
+        }
+        double[] target = Utils.cartesianToPolar(sumx, sumy);
+        FireLog.log("target velocity", target[0]);
+        FireLog.log("target angle", target[1]);
+    }
+
+    /**
+     * Turns the joystick inputs into the robot heading.
+     *
+     * @param forward    the Y value of the joystick
+     * @param strafe     the X value of the joystick
+     * @param rotation   the rotation Z of the joystick
+     * @param robotAngle the current angle of the robot in radians
+     * @return an array of the robot heading
+     */
+    public static double[] getRobotHeading(double forward, double strafe, double rotation, double robotAngle) {
+        // multiplies the 2D rotation matrix by the robot heading, there by rotating the coordinate system
+        // see https://en.wikipedia.org/wiki/Rotation_matrix
+        double[][] rotationMat = {{Math.cos(robotAngle), -Math.sin(robotAngle)},
+                {Math.sin(robotAngle), Math.cos(robotAngle)}};
+        double[] speeds = Utils.matrixVectorMult(rotationMat, new double[]{forward, strafe});
+        System.out.println("forward" + forward);
+        System.out.println("strafe" + strafe);
```

Comment:
do we need this print? if we want to see the values we should use the FireLog, am I wrong?

Reply:
you are correct this was a temporary print during testing